src/modlistcontextmenu.cpp

```diff
@@ -0,0 +1,442 @@
+        childEnabled = true;
+      }
+      checkBox->setChecked(enabled ? Qt::Checked : Qt::Unchecked);
+
+      QScopedPointer<QWidgetAction> checkableAction(new QWidgetAction(targetMenu));
+      checkableAction->setDefaultWidget(checkBox.take());
+      checkableAction->setData(id);
+      targetMenu->addAction(checkableAction.take());
+
+      if (factory.hasChildren(i)) {
+        if (populate(targetMenu, factory, mod, factory.getCategoryID(i)) || enabled) {
+          targetMenu->setIcon(QIcon(":/MO/gui/resources/check.png"));
+        }
+      }
+    }
+  }
+  return childEnabled;
+}
+
+ModListPrimaryCategoryMenu::ModListPrimaryCategoryMenu(CategoryFactory& categories, ModInfo::Ptr mod, QMenu* parent)
+  : QMenu(tr("Primary Category"), parent)
+{
+  connect(this, &QMenu::aboutToShow, [=]() { populate(categories, mod); });
+}
+
+void ModListPrimaryCategoryMenu::populate(const CategoryFactory& factory, ModInfo::Ptr mod)
+{
+  clear();
+  const std::set<int>& categories = mod->getCategories();
+  for (int categoryID : categories) {
+    int catIdx = factory.getCategoryIndex(categoryID);
+    QWidgetAction* action = new QWidgetAction(this);
+    try {
+      QRadioButton* categoryBox = new QRadioButton(
+        factory.getCategoryName(catIdx).replace('&', "&&"),
+        this);
+      categoryBox->setChecked(categoryID == mod->primaryCategory());
+      action->setDefaultWidget(categoryBox);
+      action->setData(categoryID);
+    }
+    catch (const std::exception& e) {
+      log::error("failed to create category checkbox: {}", e.what());
+    }
+
+    action->setData(categoryID);
+    addAction(action);
+  }
+}
+
+int ModListPrimaryCategoryMenu::primaryCategory() const
+{
+  for (QAction* action : actions()) {
+    QWidgetAction* widgetAction = qobject_cast<QWidgetAction*>(action);
+    if (widgetAction) {
+      QRadioButton* button = qobject_cast<QRadioButton*>(widgetAction->defaultWidget());
+      if (button && button->isChecked()) {
+        return widgetAction->data().toInt();
+      }
+    }
+  }
+  return -1;
+}
+
+ModListContextMenu::ModListContextMenu(
+  const QModelIndex& index, OrganizerCore& core, CategoryFactory& categories, ModListView* view) :
+  QMenu(view)
+  , m_core(core)
+  , m_categories(categories)
+  , m_index(index.model() == view->model() ? view->indexViewToModel(index) : index)
+  , m_view(view)
+  , m_actions(view->actions())
+{
+  if (view->selectionModel()->hasSelection()) {
+    m_selected = view->indexViewToModel(view->selectionModel()->selectedRows());
+  }
+  else {
+    m_selected = { index };
+  }
+
+  ModInfo::Ptr info = ModInfo::getByIndex(index.data(ModList::IndexRole).toInt());
+
+  QMenu* allMods = new ModListGlobalContextMenu(core, view, m_index, view->topLevelWidget());
+  allMods->setTitle(tr("All Mods"));
+  addMenu(allMods);
+
+  auto viewIndex = view->indexModelToView(m_index);
+  if (view->model()->hasChildren(viewIndex)) {
+    bool expanded = view->isExpanded(viewIndex);
```

Comment:
I would put a menu separator here `addSeparator()` after the allmods menu and before the collapse/expand options.